added cdi-stereotype quickstart

ishulga opened the pull request:
I extended cdi-interceptors quickstart and added cdi-stereotype quickstart, could you please review it?

ishulga committed:
added cdi-stereotype quickstart

pmuir commented:
@ishulga I'm confused, there isn't a cdi-interceptors quickstart...

pmuir commented:
oic, earlier pull. I'll look at those.

ishulga committed:
added cdi-decorator and cdi-interceptors

sgilda commented:
@ishulga : I'm confused. The cdi-decorator and cdi-interceptor quickstarts were previously submitted in this open pull and I reviewed them there: https://github.com/jboss-jdf/jboss-as-quickstart/pull/472

I will look at the cdi-stereotype quickstart now. In the future, it might be easier if you created a separate branch for each quickstart. :-)

@pmuir: I'm not sure how we handle multiple pulls for the same quickstarts. Should we close the previous one?

sgilda commented on cdi-decorator/README.md:
```diff
@@ -0,0 +1,94 @@
+This example demonstrates the use of CDI Decorator.
+It represents a common decorator design pattern. We take a class and we wrap decorator class around it. 
+When we call the class, we always pass through the surrounding decorator class before we reach the inner class. 
+In this example, the decorator class simply changes bonus(from 100 to 200) and position(from Java Developer to Team Lead) of 'Staff' and logs a message to the server console.
```
In this example, the decorator class simply changes the staff bonus from '100' to '200' and the staff position from 'Java Developer' to 'Team Lead'. It then logs a message to the server console.

sgilda commented on cdi-interceptors/README.md:
```diff
@@ -0,0 +1,92 @@
+-----------
+The quickstart demonstrates using CDI interceptors for cross-cutting concerns such as logging and simple auditing. 
+Interceptors can be applied to any business methods or beans, simply by adding appropriate interceptor binding type annotation. The project contains EJB service that can create and retrieve object from database.
+This example demonstrates 2 interceptors: 'AuditInterceptor' and 'LoggingInterceptor'
```
You might want to use tick marks (`) instead of quotes around the interceptor names througout the README file, here and below. For example:

This example demonstrates 2 interceptors: `AuditInterceptor` and `LoggingInterceptor`.

sgilda commented on cdi-stereotype/README.md:
```diff
@@ -0,0 +1,93 @@
+
+What is it?
+-----------
+The quickstart is extention for cdi-interceptors quickstart and demonstrates using of CDI stereotype to allow developer to declare some common metadata for beans in a central place.
```
This quickstart is an extension of the `cdi-interceptors` quickstart and demonstrates how to use a CDI stereotype. Stereotypes allow a developer to declare common metadata for beans in a central place. A stereotype is an annotation, annotated @Stereotype, that packages several other annotations.

sgilda commented on cdi-stereotype/README.md:
```diff
@@ -0,0 +1,93 @@
+In this example stereotype encapsulates combination of:
+* set of interceptor bindings ('@Logging' and '@Audit')
+* specify that all beans with the stereotype are alternatives.
+
```
The above lines don't format correctly when rendered in HTML. The following should work:

In this example, the stereotype encapsulates the following :
- All beans with this stereotype inherit the following interceptor bindings: `@Logging` and `@Audit` 
- All beans with this stereotype are alternatives

sgilda commented on cdi-stereotype/README.md:
```diff
@@ -0,0 +1,93 @@
+* set of interceptor bindings ('@Logging' and '@Audit')
+* specify that all beans with the stereotype are alternatives.
+
+A stereotype is an annotation, annotated @Stereotype, that packages several other annotations.
```
I would put this line prior to the section above.

sgilda commented on cdi-stereotype/README.md:
```diff
@@ -0,0 +1,93 @@
+
+The application will be running at the following URL <http://localhost:8080/jboss-as-cdi-stereotype>
+
+You can now comment out classes in the WEB-INF/beans.xlm file to disable one or both of the interceptors or alternative stereotype and view the results.
```
Nice job explaining how to see the different results!

sgilda commented:
@pmuir, @LightGuard : Do you want to do a code review?

LightGuard commented on cdi-interceptors/src/main/java/org/jboss/as/quickstarts/cdi/interceptor/LoggingInterceptor.java:
```diff
@@ -0,0 +1,47 @@
+    @AroundInvoke
+    public Object aroundInvoke(InvocationContext ic) throws Exception {
+        String methodName = ic.getMethod().getName();
+        logger.info("Executing " + ic.getTarget().getClass().getSimpleName() + "." + methodName + " method");
```
I think the audit listener should be doing something else, to me it would be too easy to miss the difference between the log for logging and the log for audit.

LightGuard commented on cdi-interceptors/src/main/java/org/jboss/as/quickstarts/cdi/util/Resources.java:
```diff
@@ -0,0 +1,58 @@
+    @SuppressWarnings("unused")
+    @Produces
+    @PersistenceContext
+    private EntityManager em;
```
Unless I'm mistaken we're typically producing EMs as RequestScoped

ishulga replied:
Typically in quickstarts default scope is used, so left it.

sgilda replied:
I just checked greeter (greeter/src/main/java/org/jboss/as/quickstarts/greeter/Resources.java) and kitchensink (kitchensink/src/main/java/org/jboss/as/quickstarts/kitchensink/util/Resources.java) and they both seem to use the default scope, so it looks like @ishulga is correct.

LightGuard commented on cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/service/History.java:
```diff
@@ -0,0 +1,37 @@
+public class History {
+    private static List<String> items = new ArrayList<String>();
+
+    public static List<String> getItems() {
```
I think a better name for these is in order. We already have an Item entity. Without looking at the return type for these one could easily think they'd be getting back the list of Item instances instead of the audit history.

LightGuard commented:
As we don't have any other quickstarts covering these areas, an Arquillian test or two to demonstrate how one would test these concepts would be a good idea.

ishulga committed:
Fixed README.md files and added tests

ishulga commented:
Hi, @sgilda thanks for comments, I fixed them. @LightGuard I added arquillian tests for cdi-interceptor quickstart. Could you review them? This tests are will be almost the same for these three quickstarts, should I add tests to each of them?

LightGuard commented:
There's no mention of the tests or how to run them in the interceptor README. 

As for the others, you could reference the tests in the README files if users are interested. 
—
Sent from Mailbox for iPhone

On Tue, Apr 16, 2013 at 6:46 AM, Ievgen Shulga notifications@github.com
wrote:

> I extended cdi-interceptors quickstart and added cdi-stereotype quickstart, could you please review it?
> You can merge this Pull Request by running:
>   git pull https://github.com/ishulga/jboss-as-quickstart master
> Or you can view, comment on it, or merge it online at:
>   https://github.com/jboss-jdf/jboss-as-quickstart/pull/494
> -- Commit Summary --
> - added cdi-stereotype quickstart
>   -- File Changes --
>   A cdi-stereotype/README.md (93)
>   A cdi-stereotype/pom.xml (164)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/interceptor/Audit.java (33)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/interceptor/AuditInterceptor.java (46)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/interceptor/Logging.java (33)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/interceptor/LoggingInterceptor.java (47)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/interceptor/ServiceStereotype.java (44)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/service/History.java (37)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/service/Item.java (54)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/service/ItemAlternativeServiceBean.java (50)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/service/ItemBean.java (89)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/service/ItemService.java (26)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/service/ItemServiceBean.java (45)
>   A cdi-stereotype/src/main/java/org/jboss/as/quickstarts/cdi/util/Resources.java (58)
>   A cdi-stereotype/src/main/resources/META-INF/persistence.xml (32)
>   A cdi-stereotype/src/main/webapp/WEB-INF/beans.xml (29)
>   A cdi-stereotype/src/main/webapp/WEB-INF/faces-config.xml (22)
>   A cdi-stereotype/src/main/webapp/WEB-INF/stereotype-quickstart-ds.xml (32)
>   A cdi-stereotype/src/main/webapp/index.html (23)
>   A cdi-stereotype/src/main/webapp/item.xhtml (51)
>   M pom.xml (1)
>   -- Patch Links --
>   https://github.com/jboss-jdf/jboss-as-quickstart/pull/494.patch
>   https://github.com/jboss-jdf/jboss-as-quickstart/pull/494.diff

sgilda commented:
We have standard verbiage for the Arquillian tests. You can copy the text from the kitchensink/README.md file:

## Run the Arquillian Tests 

This quickstart provides Arquillian tests. By default, these tests are configured to be skipped as Arquillian tests require the use of a container. 

_NOTE: The following commands assume you have configured your Maven user settings. If you have not, you must include Maven setting arguments on the command line. See [Run the Arquillian Tests](../README.md#arquilliantests) for complete instructions and additional options._
1. Make sure you have started the JBoss Server as described above.
2. Open a command line and navigate to the root directory of this quickstart.
3. Type the following command to run the test goal with the following profile activated:
   
   ```
   mvn clean test -Parq-jbossas-remote 
   ```

sgilda commented:
@ishulga : Please let me know when this is ready for another review. :-)

ishulga committed:
added test description in README.md

ishulga commented:
@sgilda I think it is ready. Thanks.

sgilda commented on cdi-interceptors/README.md:
```diff
@@ -0,0 +1,106 @@
+Interceptors can be applied to any business methods or beans, simply by adding appropriate interceptor binding type annotation. The project contains EJB service that can create and retrieve object from database.
+This example demonstrates 2 interceptors: `AuditInterceptor` and `LoggingInterceptor`
+
+The quickstart defines the `@Audit` and `@Logging` interceptor binding types. The `AuditInterceptor` and `LoggingInterceptor` classes are annotated with the binding type annotation and contain a method annotated `@AroundInvoke`. If the interceptor is enabled, this method will be called when the intercepted methods are invoked. In the `ItemServiceBean` bean, notice the `create()`and `getList()` me ...
```
There is a tick mark in rather than apostrophe in "bean's" and it throws off the html rendering. for the rest of this sentence :-)

sgilda commented on cdi-interceptors/README.md:
```diff
@@ -0,0 +1,106 @@
+If you have not yet done so, you must [Configure Maven](../README.md#mavenconfiguration) before testing the quickstarts.
+
+
+Run the Arquillian Tests 
```
To be consistent with the other quickstarts, would you mind moving this whole section after "Undeploy the Archive" and before "Run the Quickstart in JBoss Developer Studio or Eclipse"?

sgilda commented on cdi-stereotype/README.md:
```diff
@@ -0,0 +1,95 @@
+* Comment the '<stereotype>org.jboss.as.quickstarts.cdi.interceptor.ServiceStereotype</stereotype>' and you no longer see ItemAlternativeServiceBean implementation invoked.
+
+In this quickstart, in order to switch back to the default implementation, 
+uncomment the 'interceptors' and '<stereotype>' block in the WEB-INF/beans.xml file and redeploy the quickstart.
```
This line doesn't render correctly in HTML. You need to use tick marks instead of apostrophe. Do you want the tag elements around the "interceptors"  and the file name within tick marks as well? 

For example:

uncomment the `<interceptors>` and `<stereotype>` block in the `WEB-INF/beans.xml` file and redeploy the quickstart.

sgilda commented:
@ishulga : Were you going to address this comment by @LightGuard above: "I think a better name for these is in order. We already have an Item entity. Without looking at the return type for these one could easily think they'd be getting back the list of Item instances instead of the audit history." ?

Also, when I test the cdi-stereotype as delivered, I only see my entries displayed in the list in the browser. I don't see anything to indicate auditing or logging in the browser page or in the server log.  When I comment out one or more entries in the beans.xml file, see no difference. Am I missing something?

ishulga commented:
@sgilda What version of server are you using? I mentioned in system requirements that this quickstart doesn't work on JBoss 7.1.1 since it uses old weld library, where this bug was not fixed. To run it successfully weld libraries should be upgraded or quickstart should be deployed at JBoss EAP 6.01 or higher.

sgilda commented:
@ishulga : Sorry about that. Too many reviews. I forgot this one required a newer version of the server.
Please let me know when you've made the updates suggested above. I'll try to remember to use AS 7.2. :-)

ishulga committed:
Fixed README.md and History name.

ishulga committed:
Fixed AuditInterceptor comment.

ishulga commented:
@sgilda I fixed comments above. Could you review?

sgilda commented:
Sure thing. I'll look at it next. Thanks!

sgilda commented:
Looks good!
@LightGuard : Do you have any further comments? If not, I think this is good to merge.

LightGuard commented:
I'm good.

sgilda commented:
Squashed and merged.
